src/main.js

```diff
@@ -0,0 +1,211 @@
+import md5 from 'md5';
+import { Base64 } from 'js-base64';
+
+import {
+  VERSION,
+  DOMAIN_REGEX,
+  DEFAULTS,
+  DPR_QUALITIES
+} from './constants.js';
+
+import {
+  validateRange,
+  validateWidths,
+  validateAndDestructureOptions,
+  validateVariableQuality,
+  validateWidthTolerance
+} from './validators.js';
+
+export default class ImgixClient {
+  constructor(opts = {}) {
+    let settings = Object.assign({}, DEFAULTS);
+    this.settings = Object.assign(settings, opts);
+    // a cache to store memoized srcset width-pairs
+    this.targetWidthsCache = {};
+    if (typeof this.settings.domain != "string") {
+      throw new Error('ImgixClient must be passed a valid string domain');
+    }
+
+    if (DOMAIN_REGEX.exec(this.settings.domain) == null) {
+      throw new Error(
+        'Domain must be passed in as fully-qualified ' +
+        'domain name and should not include a protocol or any path ' +
+        'element, i.e. "example.imgix.net".');
+    }
+
+    if (this.settings.includeLibraryParam) {
+      this.settings.libraryParam = "js-" + VERSION;
+    }
+
+    this.settings.urlPrefix = this.settings.useHTTPS ? 'https://' : 'http://';
+  }
+
+  buildURL(path = '', params = {}) {
```

Comment:
Oh dang! Okay, let me fix that. Question: in an ideal world which would we have gone with `null` or `{}`? (this is mostly for my benefit/learning)

Reply:
I don't really understand the question here - if I were calling this and didn't want to pass any parameters, I would use either `client.buildURL(<url>)` or `client.buildURL(<url>, {})`. It's in the first case that params would be undefined. I almost never explicitly use `null`, but unfortunately we can't assume that others don't #billiondollarmistake

Reply:
Yeah, that pretty much answers my question. I'll switch this back to it's previous state (allowing `null`). My original reasoning for using a default parameter value here was so that this works `client.buildURL(<url>)` as well as this as well as `client.buildURL(<url>, {...})`. This is to ensure that `params` is always at least an empty object (i.e. well-defined) and to eliminate the null check in the previous iteration:

<img width="488" alt="Screen Shot 2020-12-16 at 4 22 26 PM" src="https://user-images.githubusercontent.com/22702087/102422110-e4cc6700-3fba-11eb-979e-98063ede4802.png">